README.md
# Scale Lidar Toolkit

Welcome to Scale's Lidar Toolkit.

The goal of this repository is to provide you with a set of helper functions, examples, and templates to ease the process of transforming your lidar scenes into the Scale API format.

## Documentation:

**Toolkit Documentation:** [documentation here](https://scaleapi.github.io/scale-lidar-toolkit/intro.html).

We've documented the [step by step process of using this toolkit with pictures and extensive explations](https://scaleapi.github.io/scale-lidar-toolkit/intro.html), using the file [`lidar_template.py`](https://github.com/scaleapi/scale-lidar-toolkit/blob/master/scale_format/scale-example.py) as a base.

**Scale API Documentation:**

This toolkit aids you in getting your data from it's raw format to submitted Scale Tasks ready to be worked on. We have the [Scale format documented in full in our 3d documentation](https://docs.scale.com/reference/lidar-annotation-annotation). We'd encourage you to use this as a reference for what's possible and what the different options are.

## Setup

First things first, ensure you have Pip and Python installed.

Next,

- Run `pip3 install -r requirements.txt` (to install all the dependencies; `pip` works equally well to `pip3`)

- Grab your API key from your Scale Dashboard, we'll want to set that as an enviroment variable (`SCALE_API_KEY`) in the command line or in a file

- That's it, to pass in your API key and run any script, it will look something like `SCALE_API_KEY=live_xxxx python3 <script name>`


### Optional Setup:

We recommend using AWS S3 to host your Lidar files.

If you plan to use S3, ensure you have the the AWS CLI installed by doing `aws --version` on the terminal

If you need to install it still, you can check out AWS's guide to [Manual AWS-CLI installation](https://docs.aws.amazon.com/cli/latest/userguide/install-cliv2.html)

The Lidar toolkit uses [boto3](https://boto3.amazonaws.com/) to upload files to s3, this will use aws credentials stored in `~/.aws/credentials.json` (profile default). Here's an example of what this file should look like:
```
[default]
aws_access_key_id = XFSERSDFSS...
aws_secret_access_key = Asjwje1A.....
```
Do ensure your AWS credentials have permissions to upload files to the `BUCKET` specified at the top of the [script](https://github.com/scaleapi/scale-lidar-toolkit/blob/master/lidar_template.py#L18).

### Scale File Upload:

Now the lidar toolkit support the new feature Scale File Upload.

More information about [Scale File Upload](https://docs.scale.com/reference/file-overview)

Usage:
```
scene.scale_file_upload('project_name')
```

### New release

## [0.2.0] - 2022-07-04 - scale-lidar-io-debug

### Added

- Add support for reading bs4 files


CHANGELOG.md
# Changelog

All notable changes to packages scale-lidar-io-debug and scale-lidar-io will be documented in this file.

## [0.2.0] - 2022-07-04 - scale-lidar-io-debug

### Added

- Add support for reading bs4 files

## [0.1.4] - 2022-05-04 - scale-lidar-io-debug

### Added

- Add support for bs4 and fix fisheye manual camera calibration

## [0.1.163] - 2022-02-04 - scale-lidar-io

### Fix

- Update threads in `scale_file_upload` scene method
- Update threads in `scale_s3_upload` scene method

## [0.1.162] - 2022-01-06 - scale-lidar-io

### Added

- Add threads to `scale_file_upload` scene method
- Add param to `scale_file_upload` to hide the progress bar

## [0.1.2] - 2021-10-06 - scale-lidar-io-debug

### Added

- Add param `aggregated_frames` to `preview`

## [0.1.15] - 2021-09-13

### Fixed

- No longer require S3/AWS configuration when you are not using it.
- Renamed `public_url` to `base_url`

## [0.1.14] - 2021-06-17

### Fixed

- Support Scale File Upload (alternative to S3)

## [0.1.13] - 2021-04-28

### Fixed

- Support Scale SDK +2.0

## [0.1.12] - 2021-03-08

### Fixed

- LidarImage calibrate issue

## [0.1.1] - 2021-03-05

### Fixed

- LidarImage save issue

## [0.1.1] - 2021-03-02

### Fixed

- k3-k6 distortion values issue

### Added

- Better comment on the code
- Handle better distortion values + set default to [k1,k2,p1,p2,k3,k4]
- Update example to use json instead of yaml

## [0.1.0] - 2021-02-04

### Added

- Add support for frame radar points (add_radar_point).
- Add support for frame points colors (add_colors).


pandas/pandaset-example.py
from pandaset import DataSet
import os
import sys
from io import BytesIO

import tensorflow as tf
import numpy as np
import json
import glob
from pyquaternion import Quaternion
from PIL import Image

from scale_lidar_io_debug import LidarScene, Transform

###############
## CHANGE ME ##
###############
S3_BUCKET = 'test-bucket'

def create_scene(seq, frames):
    scene = LidarScene()

    # used to remove the global values from the cameras extrinsics
    initial_pose = Transform().from_Rt(
            R=Quaternion(np.array(list(seq.lidar.poses[0]['heading'].values()))),
            t=list(seq.lidar.poses[0]['position'].values())
            )

    print(seq.camera.keys())
    for camera in seq.camera.keys():
        # create the camera extrinsics using the quaternion as rotation value and the position as translation value
        camera_extrinsics = Transform().from_Rt(
                    R=Quaternion(np.array(list(seq.camera[camera].poses[frames[0]]['heading'].values()))),
                    t=list(seq.camera[camera].poses[frames[0]]['position'].values())
                    )
        scene.get_camera(camera).calibrate(
                pose= initial_pose.inverse @ camera_extrinsics, # remove world component from camera extrinsics
                K=np.array([
                    [seq.camera[camera].intrinsics.fx, 0, seq.camera[camera].intrinsics.cx],
                    [0,seq.camera[camera].intrinsics.fy, seq.camera[camera].intrinsics.cy],
                    [0,0,1],
                    ])
                )

    # Add points to scene
    for frame in frames:
        points = seq.lidar[frame].values
        points = np.array(points[:,:4]) # keep x,y,z,i values for each point
        pose = Transform().from_Rt(
                    R=Quaternion(np.array(list(seq.lidar.poses[frame]['heading'].values()))),
                    t=list(seq.lidar.poses[frame]['position'].values())
                    )

        # Why we remove the world component from cameras and points:
        #     The lidar toolkit will apply the pose to cameras and points when it generates the output jsons,
        #     so we add all the data in ego coords + the poses in each frame
        #     and leave the ego2world conversion to the lidar toolkit
        scene.get_frame(frame).add_points(points, transform=pose.inverse)   # remove the world component from the points
        scene.get_frame(frame).transform = pose # set the pose as frame pose
        for camera in seq.camera.keys():
           scene.get_frame(frame).get_image(camera).load_pil_image(seq.camera[camera].data[frame])  # load each camera image

    scene.make_transforms_relative()    # make all frame poses relative to the first frame, this will set the first frame as position 0,0,0

    return scene

if __name__ == '__main__':
    dataset = DataSet('/data/src/pandaset')
    print(dataset.sequences())
    seq001 = dataset['001'] # load only the first sequence
    seq001.load()
    print(f"Number of frames {len(seq001.lidar[:])}")
    scene = create_scene(
        seq=seq001,
        frames=range(0,5)
    )

    # Debugging methods
    #  scene.get_projected_image("front_camera").save(f'/data/pandas.png')
    #  scene.get_frame(index=1).add_debug_lines()
    #  scene.preview()

    # Upload data to S3 bucket
    scene.s3_upload(S3_BUCKET, path='test-pandaset')
    # Create tasks/ Scale API request
    scene.create_task().publish()


scale_format/scale-example.py
import os
import numpy as np
from pyquaternion import Quaternion
import open3d as o3d
from scale_lidar_io_debug import LidarScene, Transform
import json


S3_BUCKET = 'test-bucket'
with open('template.json') as json_file:
    TEMPLATE = json.load(json_file)

def create_scene(base_path, frames):
    os.chdir(base_path)

    scene = LidarScene()
    cameras = [name for name in os.listdir("./cameras/") if os.path.isdir(os.path.join("./cameras/", name))]  # get cameras names

    for camera in cameras:
        with open(f"./cameras/{camera}/extrinsics.json") as extrinsic_file:
            camera_extrinsic = json.load(extrinsic_file)
        with open(f"./cameras/{camera}/intrinsics.json") as intrinsic_file:
            camera_intrinsic = json.load(intrinsic_file)

        scene.get_camera(camera).calibrate(
                pose=Transform().from_Rt(
                    R=Quaternion(np.array(camera_extrinsic['heading'])),
                    t=np.array(camera_extrinsic['position'])
                    ),
                K=np.array([
                    [camera_intrinsic['fx'], 0, camera_intrinsic['cx']],
                    [0,camera_intrinsic['fy'], camera_intrinsic['cy']],
                    [0,0,1],
                    ])
                # you can set here camera model if it's not 'brown_conrady' model
                # model='fisheye',
                # D= [k1, k2, p1, p2, k3],
                )

    # Add points to scene
    for frame in frames:
        # read pointcloud data in ego coordinates
        # point can be a [x,3] or [x,4] shape (intensityj)
        pcd = o3d.io.read_point_cloud(f"./pointcloud/{frame}.ply")
        with open(f"./poses/{frame}.json", 'r') as pose_file:   # load frame pose
            device_pose = json.load(pose_file)
            pose = Transform().from_Rt(
                        R=Quaternion(np.array(device_pose['heading'])),
                        t=device_pose['position']
                        )
            points = np.asarray(pcd.points)
            scene.get_frame(frame).add_points(points)
            # if points are in world coordinates you should do this
            #  scene.get_frame(frame).add_points(np.asarray(pcd.points), transform=pose.inverse)

            scene.get_frame(frame).apply_transform(pose)    # set the pose as frame pose
            for camera in cameras:  # load camera images
                scene.get_frame(frame).get_image(camera).load_file(f"./cameras/{camera}/{frame}.jpg")

    scene.make_transforms_relative()    # make all frame poses relative to the first frame, this will set the first frame as position 0,0,0

    return scene

"""
Data structure:
    - cameras:
        - camera_one:
            - [X].jpg/png (X -> frame number)
            - extrinsics.json (heading: [w,x,y,z], position: [x,y,z])
            - intrinsics.json (fx,fy,cx,cy)
        - camera_two:
            - [X].jpg/png (X -> frame number)
            - extrinsics.json (heading: [w,x,y,z], position: [x,y,z])
            - intrinsics.json (fx,fy,cx,cy)
    - pointcloud:
        - [X].ply/pcd (X -> frame number)
    - poses:
        - [X].json (X -> frame number) (heading: [w,x,y,z], position: [x,y,z])


"""
if __name__ == '__main__':
    scene = create_scene(
            'data/',
            frames=range(1,6)
            )

    # Debugging methods
    # scene.get_frame(index=1).add_debug_lines()
    # scene.preview()

    # Upload data to S3 bucket
    scene.s3_upload(S3_BUCKET, path='test-scale')
    # Create tasks/ Scale API request
    scene.create_task(TEMPLATE).publish()


scale_format/scale-example-without-S3.py
import os
import yaml
import numpy as np
from pyquaternion import Quaternion
import open3d as o3d
from scale_lidar_io_debug import LidarScene, Transform

PUBLIC_URL = "[YOUR PUBLIC HOST URL]"
SAVE_PATH = "[PATH TO STORE ZIP FILE]"
TEMPLATE = yaml.load(open('template.yml'), Loader=yaml.FullLoader)

def create_scene(base_path, frames):
    os.chdir(base_path)

    scene = LidarScene()
    cameras = [name for name in os.listdir("./cameras/") if os.path.isdir(os.path.join("./cameras/", name))]  # get cameras names

    for camera in cameras:
        with open(f"./cameras/{camera}/extrinsics.yaml", 'r') as extrinsic_file:
            camera_extrinsic = yaml.safe_load(extrinsic_file)
        with open(f"./cameras/{camera}/intrinsics.yaml", 'r') as intrinsic_file:
            camera_intrinsic = yaml.safe_load(intrinsic_file)

        scene.get_camera(camera).calibrate(
                pose=Transform().from_Rt(
                    R=Quaternion(np.array(camera_extrinsic['heading'])),
                    t=np.array(camera_extrinsic['position'])
                    ),
                K=np.array([
                    [camera_intrinsic['fx'][0], 0, camera_intrinsic['cx'][0]],
                    [0,camera_intrinsic['fy'][0], camera_intrinsic['cy'][0]],
                    [0,0,1],
                    ])
                # you can set here camera model if it's not 'brown_conrady' model
                # model='fisheye',
                # D= [k1, k2, p1, p2, k3],
                )

    # Add points to scene
    for frame in frames:
        # read pointcloud data in ego coordinates
        # point can be a [x,3] or [x,4] shape (intensityj)
        pcd = o3d.io.read_point_cloud(f"./pointcloud/{frame}.ply")
        with open(f"./poses/{frame}.yaml", 'r') as pose_file:   # load frame pose
            device_pose = yaml.safe_load(pose_file)
            pose = Transform().from_Rt(
                        R=Quaternion(np.array(device_pose['heading'])),
                        t=device_pose['position']
                        )
            points = np.asarray(pcd.points)
            scene.get_frame(frame).add_points(points)
            # if points are in world coordinates you should do this
            #  scene.get_frame(frame).add_points(np.asarray(pcd.points), transform=pose.inverse)

            scene.get_frame(frame).apply_transform(pose)    # set the pose as frame pose
            for camera in cameras:  # load camera images
                scene.get_frame(frame).get_image(camera).load_file(f"./cameras/{camera}/{frame}.jpg")

    scene.make_transforms_relative()    # make all frame poses relative to the first frame, this will set the first frame as position 0,0,0

    return scene

"""
Data structure:
    - cameras:
        - camera_one:
            - [X].jpg/png (X -> frame number)
            - extrinsics.yaml (heading: [w,x,y,z], position: [x,y,z])
            - intrinsics.yaml (fx,fy,cx,cy)
        - camera_two:
            - [X].jpg/png (X -> frame number)
            - extrinsics.yaml (heading: [w,x,y,z], position: [x,y,z])
            - intrinsics.yaml (fx,fy,cx,cy)
    - pointcloud:
        - [X].ply/pcd (X -> frame number)
    - poses:
        - [X].yaml (X -> frame number) (heading: [w,x,y,z], position: [x,y,z])
"""
if __name__ == '__main__':
    scene = create_scene(
            'data/',
            frames=range(1,6)
            )

    # Debugging methods
    #  scene.get_frame(index=1).add_debug_lines()
    #  scene.preview()

    scene.public_url = PUBLIC_URL     # this will set all the attachment using this url as public url source path ( eg. http://test.com will generate http://test.com/images_one.jpg)
    """
    First round:
        Generate the data
        **IMPORTANT** comment this on the second round
    """
    scene.save_task(SAVE_PATH)       # this will generate a zip file with all the data require for the scene, unzip this on [YOUR PUBLIC HOST URL]

    """
    Second round:
        Create the tasks after the data is on the public host url
        **IMPORTANT** uncomment this on the second round
    """
    #  # Create tasks/ Scale API request
    #  scene.create_task(TEMPLATE).publish()       # this will create the task on Scale system using the same data as before but it will reference it to the public host url


lidar_template.py
from scale_lidar_io_debug import LidarScene, Transform
import os
import json
import yaml
import numpy as np
from glob import glob

"""
Documentation for this file is in :
https://fieldengineering.gitbook.io/lidar-conversion/load-raw-lidar-data/intro
"""

###############
## CHANGE ME ##
###############
LOCAL_INPUT_PATH = ''
DATA_PATH = ''
S3_BUCKET = ''
S3_BUCKET_PATH = ''

CAMERAS = [
        'fisheye_camera',
        'front_camera'
        ]
"""
Read poses files, this will be different depending on the poses file format
"""
def read_pose(frames):
    poses = []
    for line in open('poses.txt'):
        poses.append(np.array(list(map(float, line.rstrip("\n").split(" ")))).reshape(3,4))
    if frames:
        poses = [poses[idx] for idx in frames]
    transforms = []
    for pose in poses:
        transforms.append(Transform(pose))
    return transforms

"""
Read point cloud file, with .bin extension, usually this is contains a list of points, that's why we need to reshape to -1,4 (this includes intensity)
"""
def read_lidar(frame_id):
    return np.fromfile(f'lidar:vls128/{frame_id}.bin', dtype=np.float32).reshape(-1, 4)

def load_scene(base_path, frames=None):
    os.chdir(base_path)

    """
    Definition Stage: Limit the data by frames + get source files
    """
    # read files
    files = glob('lidar:vls128/*.*', recursive=True)    # lidar files path
    frames_ids = sorted([int(frame.split("/")[-1].split(".")[0]) for frame in files])   # sort frames - this may change depending of the file names
    if frames:   # limit the number of frames
        frames_ids = [frames_ids[idx] for idx in frames]

    # New empty scene!
    scene = LidarScene()

    """
    Calibration Stage: Load global data, like poses or camera calibration
    """

    """
    Load poses.
    This will change depending on the data. You may need to just read the poses files or "normalize" it by the first frame
    """
    # Read poses
    poses = read_pose(frames)

    # Normalize poses by first frame
    offset = poses[0].inverse
    poses = [offset @ pose for pose in poses]

    # example with imu
    #  calib_imu_to_lidar = read_calib_file('calib_imu_to_lidar.txt')
    #  offset = poses[0].inverse
    #  poses = [offset @ pose for pose in poses]
    #  poses = [pose @ imu_to_lidar.inverse for pose in poses]

    """
    Add and calibrate each camera (there are multiple options to calibrate the camera, check this method on the camera class)
    """
    for camera_id in CAMERAS:
        calib = json.load(open(f'{camera_id}/camera_{camera_id}_transforms_small.txt'))
        scene.get_camera(camera_id).calibrate(
            K=np.array(calib['intrinsics']),
            pose=Transform(np.array(calib['transform'])),
        )

    #  # First example: hardcoded values
    #  scene.get_camera('rgb').calibrate(
            #  pose=Transform.from_Rt(
                #  R=np.array([
                    #  [9.99390321e-01, -2.67558875e-02, -2.24300785e-02],
                    #  [-2.24194359e-02, 6.97716781e-04, -9.99748409e-01],
                    #  [2.67648058e-02, 9.99641754e-01, 9.74394928e-05],
                    #  ]),
                #  t=np.array([-0.2319779, -0.22076198, 1.51180312])
                #  ).inverse,
            #  D=np.array([-0.15703, 0.08176, -0.00007, 0.00039, 0.00000]),
            #  K=np.array([
                #  [1882.60447, 0.0, 1230.666],
                #  [0.0, 1880.14865, 1045.725],
                #  [0.0, 0.0, 1.0]
                #  ])
            #  )

    #  # Second example: lidar to cam transformation
    #  calib_cam_to_cam = read_calib_file('calib_cam_to_cam.txt')
    #  calib_lidar_to_cam = read_calib_file('calib_lidar_to_cam.txt')
    #  lidar_to_cam = Transform.from_Rt(
            #  calib_lidar_to_cam['R'].reshape(3, 3),
            #  calib_lidar_to_cam['T'],
            #  )
    #  scene.get_camera(0).calibrate(
            #  pose=lidar_to_cam.inverse,
            #  K=calib_cam_to_cam['K_00'].reshape(3, 3) / (calib_cam_to_cam['S_00'][0] / 2896),
            #  D=calib_cam_to_cam['D_00'],
            #  )

    #  # Third example: extrinsics matrix + intrinsic focal lengths & principal point
    #  scene.get_camera(camera.name).calibrate(
            #  world_transform=Transform(camera.extrinsic.transform),
            #  fx=camera.intrinsic[0],
            #  fy=camera.intrinsic[1],
            #  cx=camera.intrinsic[2],
            #  cy=camera.intrinsic[3],
            #  )

    """
    Iteration Stage: Iterate over each frame.
    Changes made by frame should be done here
    """
    for index, frame_id in enumerate(frames_ids):
        print(f"Loading frame {frame_id}")

        # you can apply each pose in here or on the final stage
        # scene.get_frame(frame_id).apply_transform(pose)

        points = read_lidar(frame_id)
        scene.get_frame(frame_id).add_points(points)    # add points to the frame, you can also do `add_points(points, transform=[new_transform])`

        #  # alternative
        #  # if points are already in world coordinates but the cameras are not, you will need to apply the inverse of the pose to the points (turning them into ego),
        #  # so, when we create the scene, we will apply the pose to everything and points will get back to world coordinates
        #  scene.get_frame(frame_id).add_points(points, transform=pose[index].inverse)

        # Load frame image per camera
        scene.get_frame(frame_id).get_image(0).load_file(f'image_rgb/{frame_id}.png')


    """
    Final stage: Apply transformation to the full scene
    """
    # Apply poses to scene
    scene.apply_transforms(poses.values)
    # if you didn't normalize the poses by the first frame you can run this
    # scene.make_transforms_relative()

    return scene


if __name__ == '__main__':
    s = load_scene(LOCAL_INPUT_PATH, frames=range(0,20)) # load data and limit the frames to load
    s.get_frame(0).add_debug_lines()    # add lines using the camera position and heading
    #  s.get_projected_image(CAMERAS_ONE).save(f'{DATA_PATH}/debug_{frame_id}.png')
    #  s.get_frame(0).save(f'{DATA_PATH}/')

    s.preview() # preview task

    #  s.s3_upload(bucket=S3_BUCKET, path=S3_BUCKET_PATH)   # store the jsons and images on s3 bucket
    #  s.create_task().publish()    # publish the task using a template payload
